molpipeline/abstract_pipeline_elements/mol2mol/filter.py

```diff
@@ -0,0 +1,326 @@
+    @abc.abstractmethod
+    def filter_elements(
+        self,
+    ) -> Mapping[str, tuple[Optional[Union[float, int]], Optional[Union[float, int]]]]:
+        """Get filter elements as dict."""
+
+
+class BasePatternsFilter(BaseKeepMatchesFilter, abc.ABC):
+    """Filter to keep or remove molecules based on patterns."""
+
+    _patterns: dict[str, tuple[Optional[int], Optional[int]]]
+
+    def __init__(
+        self,
+        patterns: Union[
+            list[str], dict[str, Union[int, tuple[Optional[int], Optional[int]]]]
+        ],
+        keep_matches: bool = True,
+        mode: Literal["any", "all"] = "any",
+        name: Optional[str] = None,
+        n_jobs: int = 1,
+        uuid: Optional[str] = None,
+    ) -> None:
+        """Initialize BasePatternsFilter.
+
+        Parameters
+        ----------
+        patterns: Union[list[str], dict[str, Union[int, tuple[Optional[int], Optional[int]]]]]
+            List of patterns to allow in molecules.
+            Alternatively, a dictionary can be passed with patterns as keys
+            and an int for exact count or a tuple of minimum and maximum.
+        keep_matches: bool, optional (default: True)
+            If True, molecules containing the specified patterns are kept, else removed.
+        mode: Literal["any", "all"], optional (default: "any")
+            If "any", at least one of the specified patterns must be present in the molecule.
+            If "all", all of the specified patterns must be present in the molecule.
+        name: Optional[str], optional (default: None)
+            Name of the pipeline element.
+        n_jobs: int, optional (default: 1)
+            Number of parallel jobs to use.
+        uuid: str, optional (default: None)
+            Unique identifier of the pipeline element.
+        """
+        super().__init__(
+            keep_matches=keep_matches, mode=mode, name=name, n_jobs=n_jobs, uuid=uuid
+        )
+        self.patterns = patterns  # type: ignore
+
+    @property
+    def patterns(self) -> dict[str, tuple[Optional[int], Optional[int]]]:
+        """Get allowed patterns as dict."""
+        return self._patterns
+
+    @patterns.setter
+    def patterns(
+        self,
+        patterns: Union[
+            list[str], dict[str, Union[int, tuple[Optional[int], Optional[int]]]]
+        ],
+    ) -> None:
+        """Set allowed patterns as dict.
+
+        Parameters
+        ----------
+        patterns: Union[list[str], dict[str, Union[int, tuple[Optional[int], Optional[int]]]]]
+            List of patterns.
+        """
+        if isinstance(patterns, (list, set)):
+            self._patterns = {pat: (1, None) for pat in patterns}
+        else:
+            self._patterns = {
+                pat: count_value_to_tuple(count) for pat, count in patterns.items()
+            }
+
+    @property
+    def filter_elements(self) -> Mapping[str, tuple[Optional[int], Optional[int]]]:
+        """Get filter elements as dict."""
```

Comment:
Maybe add note that this was implemented to keep backwards compatibility.

Reply:
I like that you can access the patterns, descriptor names, etc. I also think "filter_names" is a suitable name for that property.

Reply:
I discussed with @JochenSiegWork that it might make sense to implement this quite generic with just filter_elements, and include a good descriptoin in the docstring

Reply:
I implemented this generically now. Please check

Reply:
Looks good!